File: sdks/unity/Runtime/Playtics.cs
// Playtics Unity SDK (Runtime)
// Minimal batching NDJSON client with gzip (optional), offline persistence, and session management.
// Drop this under Assets (e.g., Assets/Playtics/Runtime/Playtics.cs) and call Playtics.Init(...).

using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.IO.Compression;
using System.Security.Cryptography;
using System.Text;
using UnityEngine;
using UnityEngine.Networking;

namespace Playtics
{
    [Serializable]
    public class Options
    {
        public string apiKey;
        public string endpoint;  // e.g. http://localhost:8080
        public string projectId;
        public string deviceId = null; // optional override
        public int flushIntervalSec = 5;
        public int maxBatch = 50;
        public int maxQueueBytes = 512 * 1024;
        public int sessionGapSec = 30 * 60;
        public string hmacSecret = null; // not recommended in client (demo only)
        public bool debug = false;
    }

    [Serializable]
    public class Event
    {
        public string event_id;
        public string event_name;
        public string project_id;
        public string user_id;  // optional
        public string device_id;
        public string session_id; // optional
        public long ts_client; // epoch ms
        public string platform = "unity";
        public string app_version;
        public string country;
        public Dictionary<string, object> props;
    }

    public class Playtics : MonoBehaviour
    {
        public static Playtics Instance { get; private set; }

        private Options _opts;
        private string _deviceId;
        private string _userId = null;
        private string _sessionId = null;
        private long _lastActiveMs = 0;

        private readonly List<Event> _queue = new List<Event>();
        private int _queueBytes = 0;
        private bool _isFlushing = false;

        private string QueuePath => Path.Combine(Application.persistentDataPath, $"playtics_queue_{_opts.projectId}_{_deviceId}.ndjson");
        private string DevKey => $"pt_device_id_{_opts.projectId}";

        public static void Init(Options options)
        {
            if (Instance != null) return;
            var go = new GameObject("PlayticsClient");
            DontDestroyOnLoad(go);
            Instance = go.AddComponent<Playtics>();
            Instance.Configure(options);
        }

        private void Configure(Options opts)
        {
            _opts = opts;
            // device id
            _deviceId = string.IsNullOrEmpty(_opts.deviceId)
                ? (PlayerPrefs.GetString(DevKey, string.Empty))
                : _opts.deviceId;
            if (string.IsNullOrEmpty(_deviceId))
            {
                // Prefer SystemInfo.deviceUniqueIdentifier if not empty
                var duid = SystemInfo.deviceUniqueIdentifier;
                if (!string.IsNullOrEmpty(duid)) _deviceId = Hash("d_", duid);
                else _deviceId = "d_" + Guid.NewGuid().ToString("N");
                PlayerPrefs.SetString(DevKey, _deviceId);
                PlayerPrefs.Save();
            }
            LoadQueue();
            _lastActiveMs = NowMs();
            StartCoroutine(FlushLoop());
            LogDebug($"Playtics initialized. deviceId={_deviceId}, queue={_queue.Count}");
        }

        public static void SetUserId(string userId)
        {
            if (Instance == null) return;
            Instance._userId = string.IsNullOrEmpty(userId) ? null : userId;
        }

        public static string Track(string eventName, Dictionary<string, object> props = null)
        {
            if (Instance == null) return null;
            return Instance.TrackInternal(eventName, props);
        }

        public static string Expose(string exp, string variant)
        {
            return Track("experiment_exposure", new Dictionary<string, object> { { "exp", exp }, { "variant", variant } });
        }

        public static string Revenue(double amount, string currency, Dictionary<string, object> props = null)
        {
            var p = props == null ? new Dictionary<string, object>() : new Dictionary<string, object>(props);
            p["amount"] = amount; p["currency"] = currency;
            return Track("revenue", p);
        }

        public static void Flush() { if (Instance != null) Instance.StartCoroutine(Instance.FlushOnce()); }

        private string TrackInternal(string eventName, Dictionary<string, object> props)
        {
            long now = NowMs();
            RollSession(now);
            var e = new Event
            {
                event_id = UuidV7(),
                event_name = eventName,
                project_id = _opts.projectId,
                user_id = _userId,
                device_id = _deviceId,
                session_id = _sessionId,
                ts_client = now,
                platform = "unity",
                props = props
            };
            int est = EstimateSize(e);
            _queue.Add(e); _queueBytes += est; _lastActiveMs = now;
            if (_opts.debug) LogDebug($"queued {e.event_id} {e.event_name} bytes={est}");
            if (_queueBytes >= _opts.maxQueueBytes) StartCoroutine(FlushOnce());
            SaveQueue();
            return e.event_id;
        }

        private IEnumerator FlushLoop()
        {
            var wait = new WaitForSeconds(_opts.flushIntervalSec);
            while (true)
            {
                yield return wait;
                if (!_isFlushing && _queue.Count > 0)
                {
                    yield return FlushOnce();
                }
            }
        }

        private IEnumerator FlushOnce()
        {
            if (_isFlushing || _queue.Count == 0) yield break;
            _isFlushing = true;
            try
            {
                int n = Math.Min(_opts.maxBatch, _queue.Count);
                var slice = _queue.GetRange(0, n);
                string ndjson = BuildNdjson(slice);
                byte[] body = Encoding.UTF8.GetBytes(ndjson);
                bool gzOk = TryGzip(ref body);

                var url = _opts.endpoint.TrimEnd('/') + "/v1/batch";
                var req = new UnityWebRequest(url, UnityWebRequest.kHttpVerbPOST);
                req.uploadHandler = new UploadHandlerRaw(body);
                req.downloadHandler = new DownloadHandlerBuffer();
                req.SetRequestHeader("x-api-key", _opts.apiKey);
                req.SetRequestHeader("content-type", "application/x-ndjson");
                if (gzOk) req.SetRequestHeader("content-encoding", "gzip");

                if (!string.IsNullOrEmpty(_opts.hmacSecret))
                {
                    string t = (NowMs() / 1000L).ToString();
                    // HMAC must sign the uncompressed body per our contract; here we sign the original NDJSON
                    byte[] raw = Encoding.UTF8.GetBytes(ndjson);
                    string sig = HmacSha256Hex(_opts.hmacSecret, t + "." + Encoding.UTF8.GetString(raw));
                    req.SetRequestHeader("x-signature", $"t={t}, s={sig}");
                }

                yield return req.SendWebRequest();
                if (req.result == UnityWebRequest.Result.Success && req.responseCode >= 200 && req.responseCode < 300)
                {
                    // drop sent slice
                    _queue.RemoveRange(0, n);
                    RecalcQueueBytes();
                    if (_opts.debug) LogDebug($"flushed {n} events ok");
                }
                else
                {
                    if (_opts.debug) LogDebug($"flush failed: {req.responseCode} {req.error}");
                    // keep queue intact; next interval will retry
                }
            }
            finally
            {
                SaveQueue();
                _isFlushing = false;
            }
        }

        private void RollSession(long nowMs)
        {
            if (string.IsNullOrEmpty(_sessionId) || nowMs - _lastActiveMs > _opts.sessionGapSec * 1000L)
            {
                _sessionId = UuidV7();
            }
        }

        private void LoadQueue()
        {
            try
            {
                if (!File.Exists(QueuePath)) return;
                var lines = File.ReadAllLines(QueuePath);
                foreach (var line in lines)
                {
                    // best-effort: we do not parse back to objects; we just ignore persisted lines
                    // Alternatively, parse minimal fields to requeue. Keep simple for skeleton.
                }
                // Truncate persisted queue after load
                File.WriteAllText(QueuePath, string.Empty);
            }
            catch { /* ignore */ }
        }

        private void SaveQueue()
        {
            try
            {
                var sb = new StringBuilder(_queue.Count * 128);
                foreach (var e in _queue) { sb.AppendLine(ToJson(e)); }
                File.WriteAllText(QueuePath, sb.ToString());
            }
            catch { /* ignore */ }
        }

        private void RecalcQueueBytes()
        {
            _queueBytes = 0;
            foreach (var e in _queue) _queueBytes += EstimateSize(e);
        }

        private static int EstimateSize(Event e) => Encoding.UTF8.GetByteCount(ToJson(e)) + 1;

        private static string BuildNdjson(List<Event> events)
        {
            var sb = new StringBuilder(events.Count * 128);
            for (int i = 0; i < events.Count; i++)
            {
                sb.Append(ToJson(events[i]));
                if (i < events.Count - 1) sb.Append('\n');
            }
            return sb.ToString();
        }

        private static string ToJson(Event e)
        {
            var sb = new StringBuilder(256);
            sb.Append('{');
            JField(sb, "event_id", e.event_id);
            JField(sb, "event_name", e.event_name);
            JField(sb, "project_id", e.project_id);
            if (!string.IsNullOrEmpty(e.user_id)) JField(sb, "user_id", e.user_id);
            JField(sb, "device_id", e.device_id);
            if (!string.IsNullOrEmpty(e.session_id)) JField(sb, "session_id", e.session_id);
            JField(sb, "ts_client", e.ts_client);
            if (!string.IsNullOrEmpty(e.platform)) JField(sb, "platform", e.platform);
            if (!string.IsNullOrEmpty(e.app_version)) JField(sb, "app_version", e.app_version);
            if (!string.IsNullOrEmpty(e.country)) JField(sb, "country", e.country);
            if (e.props != null && e.props.Count > 0) JObject(sb, "props", e.props);
            if (sb[sb.Length - 1] == ',') sb.Length -= 1; // trim last comma
            sb.Append('}');
            return sb.ToString();
        }

        private static void JField(StringBuilder sb, string k, string v)
        {
            sb.Append('"').Append(JEscape(k)).Append('"').Append(':')
              .Append('"').Append(JEscape(v)).Append('"').Append(',');
        }
        private static void JField(StringBuilder sb, string k, long v)
        {
            sb.Append('"').Append(JEscape(k)).Append('"').Append(':').Append(v).Append(',');
        }
        private static void JObject(StringBuilder sb, string k, Dictionary<string, object> map, int depth = 0)
        {
            if (depth > 2) return; // clamp
            sb.Append('"').Append(JEscape(k)).Append('"').Append(':');
            sb.Append('{');
            int count = 0;
            foreach (var kv in map)
            {
                if (count >= 50) break;
                sb.Append('"').Append(JEscape(kv.Key)).Append('"').Append(':');
                JValue(sb, kv.Value, depth + 1);
                sb.Append(',');
                count++;
            }
            if (sb[sb.Length - 1] == ',') sb.Length -= 1;
            sb.Append('}').Append(',');
        }
        private static void JArray(StringBuilder sb, IList list, int depth)
        {
            if (depth > 2) { sb.Append("[]"); return; }
            sb.Append('[');
            int lim = Math.Min(50, list.Count);
            for (int i = 0; i < lim; i++)
            {
                JValue(sb, list[i], depth + 1);
                sb.Append(',');
            }
            if (sb[sb.Length - 1] == ',') sb.Length -= 1;
            sb.Append(']');
        }
        private static void JValue(StringBuilder sb, object v, int depth)
        {
            if (v == null) { sb.Append("null"); return; }
            switch (v)
            {
                case string s:
                    sb.Append('"').Append(JEscape(s)).Append('"'); break;
                case bool b:
                    sb.Append(b ? "true" : "false"); break;
                case int or long or float or double or decimal:
                    sb.Append(Convert.ToString(v, System.Globalization.CultureInfo.InvariantCulture)); break;
                case Dictionary<string, object> m:
                    JObject(sb, null, m, depth); break; // key null means direct object (only used in nested)
                case IList list:
                    JArray(sb, list, depth); break;
                default:
                    sb.Append('"').Append(JEscape(v.ToString())).Append('"'); break;
            }
        }
        private static string JEscape(string s) => s.Replace("\\", "\\\\").Replace("\"", "\\\"");

        private static bool TryGzip(ref byte[] body)
        {
            try
            {
                using var ms = new MemoryStream();
                using (var gz = new GZipStream(ms, CompressionLevel.Fastest, leaveOpen: true))
                {
                    gz.Write(body, 0, body.Length);
                }
                body = ms.ToArray();
                return true;
            }
            catch { return false; }
        }

        private static string HmacSha256Hex(string secret, string message)
        {
            try
            {
                using var h = new HMACSHA256(Encoding.UTF8.GetBytes(secret));
                var bytes = h.ComputeHash(Encoding.UTF8.GetBytes(message));
                var sb = new StringBuilder(bytes.Length * 2);
                foreach (var b in bytes) sb.AppendFormat("{0:x2}", b);
                return sb.ToString();
            }
            catch { return null; }
        }

        private static long NowMs() => DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        private static string Hash(string prefix, string v)
        {
            using var sha = SHA1.Create();
            var bytes = sha.ComputeHash(Encoding.UTF8.GetBytes(v));
            var sb = new StringBuilder(prefix);
            for (int i = 0; i < 12; i++) sb.Append(bytes[i].ToString("x2"));
            return sb.ToString();
        }
        private static string UuidV7()
        {
            long ms = NowMs();
            string ts = ms.ToString("x").PadLeft(12, '0');
            string rnd = Guid.NewGuid().ToString("N");
            string hex = ts + rnd.Substring(0, 20);
            // xxxx-xxxx-7xxx-xxxx-xxxxxxxxxxxx
            return hex.Substring(0, 8) + "-" + hex.Substring(8, 12 - 8) + "-7" + hex.Substring(13, 16 - 13) + "-" + hex.Substring(16, 20 - 16) + "-" + hex.Substring(20, 32 - 20);
        }

        private static void LogDebug(string msg) { Debug.Log("[Playtics] " + msg); }

        // ===== Experiments helpers =====
        public static string AssignVariant(string expId, string salt, System.Collections.Generic.List<System.Tuple<string,int>> variants, string key)
        {
            if (variants == null || variants.Count == 0) return "A";
            int sum = 0; foreach (var v in variants) sum += v.Item2 > 0 ? v.Item2 : 1;
            uint h = Hash32(expId+":"+(salt??"")+":"+key);
            int r = (int)(h % (uint)sum);
            int acc = 0; foreach (var v in variants) { acc += v.Item2>0?v.Item2:1; if (r < acc) return v.Item1; }
            return variants[0].Item1;
        }
        private static uint Hash32(string s)
        {
            uint h = 0x811c9dc5;
            foreach (char ch in s) { h ^= (uint)ch; h += (h<<1)+(h<<4)+(h<<7)+(h<<8)+(h<<24); }
            return h;
        }
    }
}

File: sdks/android/playtics-android/src/main/java/io/playtics/android/Playtics.kt
package io.playtics.android

import android.content.Context
import android.content.SharedPreferences
import android.util.Base64
import java.io.ByteArrayOutputStream
import java.util.UUID
import java.util.Timer
import java.util.TimerTask
import java.util.concurrent.atomic.AtomicBoolean
import java.util.zip.GZIPOutputStream
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody

/**
 * Minimal Android SDK that mirrors the Web SDK behavior: batching, NDJSON, optional gzip & HMAC,
 * offline persistence (SharedPreferences), and simple session handling.
 */
class Playtics(private val ctx: Context, private val opts: Options) {
  data class Options(
    val apiKey: String,
    val endpoint: String,
    val projectId: String,
    val deviceId: String? = null,
    val flushIntervalMs: Long = 5000,
    val maxBatch: Int = 50,
    val maxQueueBytes: Int = 512_000,
    val sessionGapMs: Long = 30*60*1000,
    val hmacSecret: String? = null, // Caution: embedding secret in client is not recommended for production.
    val debug: Boolean = false
  )

  data class Event(
    val event_id: String,
    val event_name: String,
    val project_id: String,
    val user_id: String? = null,
    val device_id: String,
    val session_id: String? = null,
    val ts_client: Long,
    val platform: String? = "android",
    val app_version: String? = null,
    val country: String? = null,
    val props: Map<String, Any?>? = null
  )

  private val client = OkHttpClient()
  private val prefs: SharedPreferences = ctx.getSharedPreferences("playtics", Context.MODE_PRIVATE)
  private val queue = ArrayList<Event>()
  private var queueBytes: Int = 0
  private val timer = Timer(true)
  private val flushing = AtomicBoolean(false)

  private var deviceId: String
  private var userId: String? = null
  private var sessionId: String? = null
  private var lastActive: Long = System.currentTimeMillis()

  init {
    deviceId = opts.deviceId ?: prefs.getString(devKey(), null) ?: genDeviceId().also {
      prefs.edit().putString(devKey(), it).apply()
    }
    restoreQueue()
    ensureTimer()
  }

  fun setUserId(uid: String?) { userId = uid }
  fun setUserProps(props: Map<String, Any?>) { /* optional; merge into props per event if needed */ }

  fun track(eventName: String, props: Map<String, Any?>? = null): String {
    val now = System.currentTimeMillis()
    rollSession(now)
    val evt = Event(
      event_id = uuidv7(),
      event_name = eventName,
      project_id = opts.projectId,
      user_id = userId,
      device_id = deviceId,
      session_id = sessionId,
      ts_client = now,
      platform = "android",
      props = props
    )
    val est = evt.estimateSize()
    queue.add(evt)
    queueBytes += est
    lastActive = now
    if (opts.debug) android.util.Log.d("Playtics", "queued ${evt.event_id} ${evt.event_name}")
    if (queueBytes >= opts.maxQueueBytes) flush()
    persistQueue()
    return evt.event_id
  }

  fun expose(exp: String, variant: String) = track("experiment_exposure", mapOf("exp" to exp, "variant" to variant))
  fun revenue(amount: Number, currency: String, props: Map<String, Any?>? = null) =
    track("revenue", mapOf("amount" to amount, "currency" to currency) + (props ?: emptyMap()))

  @Synchronized fun flush() {
    if (queue.isEmpty() || !flushing.compareAndSet(false, true)) return
    try {
      val batch = ArrayList<Event>()
      val n = minOf(opts.maxBatch, queue.size)
      for (i in 0 until n) batch.add(queue.removeAt(0))
      recalcQueueBytes()
      send(batch)
    } finally {
      flushing.set(false)
      persistQueue()
    }
  }

  fun shutdown() {
    try { timer.cancel() } catch (_: Throwable) {}
  }

  private fun ensureTimer() {
    timer.schedule(object : TimerTask() {
      override fun run() { try { flush() } catch (_: Throwable) {} }
    }, opts.flushIntervalMs, opts.flushIntervalMs)
  }

  private fun rollSession(now: Long) {
    if (sessionId == null || now - lastActive > opts.sessionGapMs) {
      sessionId = uuidv7()
    }
  }

  private fun send(batch: List<Event>) {
    if (batch.isEmpty()) return
    val ndjson = buildString(batch.size * 128) {
      val moshi = Json
      for (e in batch) {
        append(moshi.stringify(e)).append('\n')
      }
    }
    var bodyBytes = ndjson.toByteArray(Charsets.UTF_8)
    val reqBuilder = Request.Builder()
      .url(opts.endpoint.trimEnd('/') + "/v1/batch")
      .addHeader("x-api-key", opts.apiKey)
      .addHeader("content-type", "application/x-ndjson")

    // gzip if helps
    val gz = gzip(bodyBytes)
    if (gz != null) {
      bodyBytes = gz
      reqBuilder.addHeader("content-encoding", "gzip")
    }

    // optional HMAC signature (not recommended to ship secret in client)
    opts.hmacSecret?.let { secret ->
      val t = (System.currentTimeMillis() / 1000).toString()
      val msg = "$t." + String(bodyBytes, Charsets.UTF_8)
      val s = hmacSha256Hex(secret, msg)
      reqBuilder.addHeader("x-signature", "t=$t, s=$s")
    }

    val req = reqBuilder.post(RequestBody.create("application/octet-stream".toMediaType(), bodyBytes)).build()
    try {
      client.newCall(req).execute().use { resp ->
        if (!resp.isSuccessful) {
          // on failure, requeue
          queue.addAll(0, batch)
          recalcQueueBytes()
        }
      }
    } catch (e: Exception) {
      // offline or network error: requeue
      queue.addAll(0, batch)
      recalcQueueBytes()
    }
  }

  // Simple JSON without external libs
  private object Json {
    fun stringify(e: Event): String {
      val sb = StringBuilder(256)
      sb.append('{')
      field(sb, "event_id", e.event_id)
      field(sb, "event_name", e.event_name)
      field(sb, "project_id", e.project_id)
      e.user_id?.let { field(sb, "user_id", it) }
      field(sb, "device_id", e.device_id)
      e.session_id?.let { field(sb, "session_id", it) }
      field(sb, "ts_client", e.ts_client)
      e.platform?.let { field(sb, "platform", it) }
      e.app_version?.let { field(sb, "app_version", it) }
      e.country?.let { field(sb, "country", it) }
      e.props?.let { objField(sb, "props", it) }
      // remove trailing comma
      if (sb.last() == ',') sb.setLength(sb.length - 1)
      sb.append('}')
      return sb.toString()
    }

    private fun field(sb: StringBuilder, k: String, v: String) {
      sb.append('"').append(escape(k)).append('"').append(':')
        .append('"').append(escape(v)).append('"').append(',')
    }
    private fun field(sb: StringBuilder, k: String, v: Long) {
      sb.append('"').append(escape(k)).append('"').append(':').append(v).append(',')
    }
    private fun objField(sb: StringBuilder, k: String, obj: Map<String, Any?>) {
      sb.append('"').append(escape(k)).append('"').append(':').append(mapToJson(obj)).append(',')
    }
    private fun mapToJson(m: Map<String, Any?>, depth: Int = 0): String {
      if (depth > 2) return "{}"
      val sb = StringBuilder("{")
      for ((kk,vv) in m) {
        sb.append('"').append(escape(kk)).append('"').append(':').append(valueToJson(vv, depth+1)).append(',')
      }
      if (sb.last() == ',') sb.setLength(sb.length - 1)
      sb.append('}')
      return sb.toString()
    }
    private fun valueToJson(v: Any?, depth: Int): String {
      return when (v) {
        null -> "null"
        is String -> '"' + escape(v) + '"'
        is Number, is Boolean -> v.toString()
        is Map<*,*> -> mapToJson(v as Map<String, Any?>, depth)
        is List<*> -> listToJson(v as List<Any?>, depth)
        else -> '"' + escape(v.toString()) + '"'
      }
    }
    private fun listToJson(l: List<Any?>, depth: Int): String {
      if (depth > 2) return "[]"
      val sb = StringBuilder("[")
      val lim = minOf(50, l.size)
      for (i in 0 until lim) {
        sb.append(valueToJson(l[i], depth+1)).append(',')
      }
      if (sb.last() == ',') sb.setLength(sb.length - 1)
      sb.append(']')
      return sb.toString()
    }
    private fun escape(s: String): String = s.replace("\\", "\\\\").replace("\"", "\\\"")
  }

  private fun Event.estimateSize(): Int {
    // rough estimation for queue size accounting
    return Json.stringify(this).toByteArray(Charsets.UTF_8).size + 1
  }

  private fun gzip(input: ByteArray): ByteArray? {
    return try {
      val bout = ByteArrayOutputStream()
      GZIPOutputStream(bout).use { it.write(input) }
      bout.toByteArray()
    } catch (_: Throwable) { null }
  }

  private fun devKey() = "pt_device_id_${opts.projectId}"
  private fun queueKey() = "pt_queue_${opts.projectId}_${deviceId}"

  private fun persistQueue() {
    try {
      val arr = StringBuilder(queue.size * 128)
      for (e in queue) {
        arr.append(Json.stringify(e)).append('\n')
      }
      prefs.edit().putString(queueKey(), arr.toString()).apply()
    } catch (_: Throwable) {}
  }
  private fun restoreQueue() {
    try {
      val s = prefs.getString(queueKey(), null) ?: return
      val lines = s.split('\n').filter { it.isNotBlank() }
      for (ln in lines) {
        // Simple restore: we don't parse back, only drop persisted queue to avoid crash; could parse if needed.
        // To keep simple, ignore restore content for now.
      }
    } catch (_: Throwable) {}
  }

  private fun hmacSha256Hex(secret: String, message: String): String {
    val mac = Mac.getInstance("HmacSHA256")
    mac.init(SecretKeySpec(secret.toByteArray(Charsets.UTF_8), "HmacSHA256"))
    val out = mac.doFinal(message.toByteArray(Charsets.UTF_8))
    return out.joinToString("") { b -> "%02x".format(b) }
  }

  private fun genDeviceId(): String = "d_" + UUID.randomUUID().toString().replace("-", "")

  private fun uuidv7(): String {
    val t = System.currentTimeMillis()
    val ts = java.lang.Long.toHexString(t).padStart(12, '0')
    val rnd = UUID.randomUUID().toString().replace("-", "")
    val hex = ts + rnd.take(20)
    return hex.substring(0,8) + "-" + hex.substring(8,12) + "-7" + hex.substring(13,16) + "-" + hex.substring(16,20) + "-" + hex.substring(20,32)
  }

  // ===== Experiments helpers =====
  data class Variant(val name: String, val weight: Int)
  private fun hash32(s: String): Int {
    var h = 0x811c9dc5.toInt()
    for (ch in s) { h = h xor ch.code; h += (h shl 1) + (h shl 4) + (h shl 7) + (h shl 8) + (h shl 24) }
    return h ushr 0
  }
  fun assignVariant(expId: String, salt: String?, variants: List<Variant>, key: String): String {
    if (variants.isEmpty()) return "A"
    val sum = variants.sumOf { if (it.weight>0) it.weight else 1 }
    val h = (hash32("$expId:${salt?:("")}:$key") % sum + sum) % sum
    var acc = 0
    for (v in variants) { acc += if (v.weight>0) v.weight else 1; if (h < acc) return v.name }
    return variants[0].name
  }
  /** Fetch experiments config (running) from control-service. WARNING: network on caller thread. */
  fun fetchExperiments(controlEndpoint: String, projectId: String): String? {
    return try {
      val url = controlEndpoint.trimEnd('/') + "/api/config/" + projectId
      val req = okhttp3.Request.Builder().url(url).get().build()
      client.newCall(req).execute().use { resp -> if (resp.isSuccessful) resp.body?.string() else null }
    } catch (_: Throwable) { null }
  }
}
